### PR TITLE
Add git support for compatibility checker

### PR DESCRIPTION
**Pull request by mannoopj**
Add git support for schema compatibility checker. Pulls in valid schema from remote git trunk branch to check with edited schema in local branch. Adds new option for command line `verify-evolution-git` which takes in a required file name

### CODE REVIEW

File: generator/src/main/java/org/apache/kafka/message/checker/CheckerUtils.java [hunk 1]
```diff
@@ -91,4 +102,54 @@ static MessageSpec readMessageSpecFromFile(String schemaPath) {
...
         }
     }
+
+    /**
+     * Return a MessageSpec file give file contents.
+     *
+     * @param contents      The path to read the file from.
+     * @return              The MessageSpec.
+     */
+    static MessageSpec readMessageSpecFromString(String contents) {
```
Comment (cmccabe) [inline generator/src/main/java/org/apache/kafka/message/checker/CheckerUtils.java:114]: we don't need to create a file to translate a string to a MessageSpec, we can do this directly.

File: generator/src/main/java/org/apache/kafka/message/checker/CheckerUtils.java [hunk 2]
```diff
@@ -91,4 +102,54 @@ static MessageSpec readMessageSpecFromFile(String schemaPath) {
...
+    }
+
+
+    /**
+     * Read a MessageSpec file from remote git repo.
+     *
+     * @param fileCheckMetadata    The file to read from remote git repo.
+     * @return                     The file contents.
+     */
+    static String GetDataFromGit(String fileCheckMetadata) throws IOException {
```
Comment (cmccabe) [inline generator/src/main/java/org/apache/kafka/message/checker/CheckerUtils.java]: I don't understand why the parameter is named `fileCheckMetadata`, can you explain?

File: generator/src/main/java/org/apache/kafka/message/checker/CheckerUtils.java [hunk 3]
```diff
@@ -91,4 +102,54 @@ static MessageSpec readMessageSpecFromFile(String schemaPath) {
...
+        Repository repository = git.getRepository();
+        Ref head = git.getRepository().getRefDatabase().firstExactRef("refs/heads/trunk");
+
+        try (RevWalk revWalk = new RevWalk(repository)) {
+            RevCommit commit = revWalk.parseCommit(head.getObjectId());
+            RevTree tree = commit.getTree();
+            try (TreeWalk treeWalk = new TreeWalk(repository)) {
+                treeWalk.addTree(tree);
+                treeWalk.setRecursive(true);
+                treeWalk.setFilter(PathFilter.create(String.valueOf(Paths.get("metadata/src/main/resources/common/metadata/" + fileCheckMetadata))));
```
Comment (cmccabe) [inline generator/src/main/java/org/apache/kafka/message/checker/CheckerUtils.java]: please use a constant for `metadata/src/main/resources/common/metadata`

File: generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java [hunk 4]
```diff
@@ -56,6 +73,11 @@ public static void run(
         evolutionVerifierParser.addArgument("--path2", "-2").
             required(true).
             help("The final schema JSON path.");
+        Subparser evolutionGitVerifierParser = subparsers.addParser("verify-evolution-git").
+            help(" Verify that an evolution of a JSon file is valid using git. ");
```
Comment (cmccabe) [inline generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java]: `JSon` ?

File: generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java [hunk 5]
```diff
@@ -81,8 +103,18 @@ public static void run(
                         ", and path2: " + path2);
                 break;
             }
+            case "verify-evolution-git": {
+                String fileCheckMetadata = namespace.getString("file");
+                String gitContent = GetDataFromGit(fileCheckMetadata);
+                EvolutionVerifier verifier = new EvolutionVerifier(
+                        CheckerUtils.readMessageSpecFromFile(Paths.get("").toAbsolutePath().getParent() + "/metadata/src/main/resources/common/metadata/" + fileCheckMetadata),
```
Comment (cmccabe) [inline generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java]: in general I don't understand what the strategy is here for paths. if you are using a relative path, you need to document what directory we need to be in. how are we locating the git repository root?

File: tools/src/main/java/org/apache/kafka/tools/ClusterTool.java [hunk 6]
```diff
@@ -37,6 +37,8 @@
 
 import static net.sourceforge.argparse4j.impl.Arguments.store;
 
+import org.eclipse.jgit.api.Git;
```
Comment (cmccabe) [inline tools/src/main/java/org/apache/kafka/tools/ClusterTool.java]: why is this import needed here?

File: generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java (hunk 5 above)
Comment (mannoopj) [reply generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java]: Ok yeah I added documentation in the help output stating that this tool is expected to be run from the `generator` directory. That was the assumption I was working under. Getting the git root repo is done in the `GetDataFromGit` function in `CheckerUtil.java`

File: tools/src/main/java/org/apache/kafka/tools/ClusterTool.java (hunk 6 above)
Comment (mannoopj) [reply tools/src/main/java/org/apache/kafka/tools/ClusterTool.java]: not needed

File: generator/src/main/java/org/apache/kafka/message/checker/CheckerUtils.java (hunk 2 above)
Comment (mannoopj) [reply generator/src/main/java/org/apache/kafka/message/checker/CheckerUtils.java]: Was to check the metadata schema file, the variable is renamed now

File: generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java (hunk 5 above)
Comment (cmccabe) [reply generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java]: If we know where the git repo root is, why can't we just change to that directory rather than requiring a specific current working directory?

File: checkstyle/suppressions.xml [hunk 7]
```diff
@@ -32,6 +32,8 @@
               files="(ApiMessageType|FieldSpec|MessageDataGenerator|KafkaConsumerTest).java"/>
     <suppress checks="MethodLength"
               files="(FieldSpec|MessageDataGenerator).java"/>
+    <suppress checks="ImportControl"
```
Comment (cmccabe) [inline checkstyle/suppressions.xml]: why can we not just list `org.eclipse.jgit` as allowed?

Comment (mannoopj) [reply checkstyle/suppressions.xml]: Was having trouble testing to so added it in the suppressions. Figured out I was having a local issue so this is just added to the allowed import controls now.

File: generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java (hunk 5 above)
Comment (mannoopj) [reply generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java]: I changed to code so it works anywhere within the kafka directory. Im trying to get the absolute path to the kafka directory so that I can use that to git repo through jgit

**Review by cmccabe — APPROVED** [review]
LGTM

**Comment by cmccabe** [comment]
@mannoopj : there are checkstyle issues, can you fix?

**Review by cmccabe — APPROVED** [review]
LGTM

File: generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java [hunk 8]
```diff
@@ -81,6 +90,23 @@ public static void run(
                         ", and path2: " + path2);
                 break;
             }
+            case "verify-evolution-git": {
+                String filePath = "/metadata/src/main/resources/common/metadata/" + namespace.getString("file");
+                Path rootKafkaDirectory = Paths.get("").toAbsolutePath();
+                while (!rootKafkaDirectory.endsWith("kafka")) {
```
Comment (chia7712) [inline generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java:96]: this condition depends on the project folder name. What we're actually looking for is the Git folder, so maybe we could change the condition to `while (!new File(rootKafkaDirectory.toFile(), ".git").exists())`, which would allow us to use a different folder name.

**Review by chia7712 — COMMENTED** [review]
Sorry for late review. This PR is great but I have one small suggestion. PTAL

File: generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java (hunk 8 above)
Comment (chia7712) [reply generator/src/main/java/org/apache/kafka/message/checker/MetadataSchemaCheckerTool.java:96]: I have filed https://issues.apache.org/jira/browse/KAFKA-18128 to fix it and another failed test related to this PR.
